dbms/src/Storages/DeltaMerge/LocalIndexerScheduler.cpp

```diff
@@ -0,0 +1,409 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <Storages/DeltaMerge/LocalIndexerScheduler.h>
+#include <common/logger_useful.h>
+
+namespace DB::DM
+{
+
+LocalIndexerScheduler::LocalIndexerScheduler(const Options & options)
+    : logger(Logger::get())
+    , pool(std::make_unique<ThreadPool>(options.pool_size, options.pool_size, options.pool_size + 1))
+    , pool_max_memory_limit(options.memory_limit)
+    , pool_current_memory(0)
+{
+    // QueueSize = PoolSize+1, because our scheduler will try to schedule next task
+    // right before the current task is finished.
+
+    LOG_INFO(
+        logger,
+        "Initialized LocalIndexerScheduler, pool_size={}, memory_limit_mb={:.1f}",
+        options.pool_size,
+        static_cast<double>(options.memory_limit) / 1024 / 1024);
+
+    if (options.auto_start)
+        start();
+}
+
+LocalIndexerScheduler::~LocalIndexerScheduler()
```

Comment:
Is it possible that a InternalTask holds a strong reference to LocalIndexerScheduler?

Reply:
Is it for canceling the task quickly?

Reply:
If the InternalTask is the last one to hold a strong reference to LocalIndexerScheduler, then deadlock could happen

Reply:
Oh, get it. InternalTask will never hold a strong reference to LocalIndexerScheduler.